crates/cxx-qt-lib/src/core/qurl.rs

```diff
@@ -63,21 +196,190 @@ pub struct QUrl {
+        ffi::qurl_authority(self)
+    }
+
+    /// Returns the name of the file, excluding the directory path.
+    ///
+    /// Note that, if this QUrl object is given a path ending in a slash, the name of the file is considered empty.
+    ///
+    /// If the path doesn't contain any slash, it is fully returned as the fileName.
+    pub fn file_name(&self) -> ffi::QString {
+        ffi::qurl_file_name(self)
+    }
+
+    /// Returns the fragment of the URL.
+    pub fn fragment(&self) -> Option<ffi::QString> {
+        if self.has_fragment() {
+            Some(self.fragment_or_default())
+        } else {
+            None
+        }
+    }
+
+    /// Returns the fragment of the URL if it is defined; otherwise an empty string is returned.
+    pub fn fragment_or_default(&self) -> ffi::QString {
+        ffi::qurl_fragment(self)
+    }
+
+    /// Parses input and returns the corresponding QUrl. input is assumed to be in encoded form, containing only ASCII characters.
+    pub fn from_encoded(input: &ffi::QByteArray) -> Self {
+        ffi::qurl_from_encoded(input)
+    }
+
+    /// Returns a QUrl representation of localFile, interpreted as a local file.
+    /// This function accepts paths separated by slashes as well as the native separator for this platform.
+    pub fn from_local_file(local_file: &ffi::QString) -> Self {
+        ffi::qurl_from_local_file(local_file)
+    }
+
+    /// Returns a decoded copy of input. input is first decoded from percent encoding,
+    /// then converted from UTF-8 to unicode.
+    pub fn from_percent_encoding(input: &ffi::QByteArray) -> ffi::QString {
+        ffi::qurl_from_percent_encoding(input)
+    }
+
+    /// Returns a valid URL from a user supplied userInput string if one can be deduced.
+    /// In the case that is not possible, an invalid QUrl() is returned.
+    pub fn from_user_input(user_input: &ffi::QString, working_directory: &ffi::QString) -> Self {
+        ffi::qurl_from_user_input(user_input, working_directory)
+    }
+
+    /// Returns the host of the URL if it is defined; otherwise an empty string is returned.
+    pub fn host_or_default(&self) -> ffi::QString {
+        ffi::qurl_host(self)
+    }
+
+    /// Returns the current whitelist of top-level domains that are allowed to have non-ASCII characters in their compositions.
+    pub fn idn_whitelist() -> ffi::QStringList {
+        ffi::qurl_idn_whitelist()
+    }
+
+    /// Returns the password of the URL if it is defined; otherwise an empty string is returned.
+    pub fn password_or_default(&self) -> ffi::QString {
+        ffi::qurl_password(self)
+    }
+
+    /// Returns the path of the URL.
+    pub fn path(&self) -> ffi::QString {
+        ffi::qurl_path(self)
+    }
+
+    /// Returns the query string of the URL if there's a query string
+    pub fn query(&self) -> Option<ffi::QString> {
+        if self.has_fragment() {
+            Some(self.fragment_or_default())
+        } else {
+            None
+        }
+    }
```

Comment:
``` diff
- This should use `has_query` and `query_or_default`!
```

Reply:
Oops :-)

Reply:
No worries, just made it red to stand out in the bazillion other comments :wink: 
This one's actually important :D

Reply:
Done